File: openless-all/app/src/lib/ipc/billing.ts
// Billing IPC bindings for ZeroType.
import { invokeOrMock } from "./shared"

export interface QuotaStatus {
  charsUsed: number
  charsLimit: number
  charsRemaining: number
  weekId: string
  isOverLimit: boolean
  percentUsed: number
}

export function getQuotaStatus(): Promise<QuotaStatus> {
  return invokeOrMock("get_quota_status", undefined, () => ({
    charsUsed: 2500,
    charsLimit: 10000,
    charsRemaining: 7500,
    weekId: "2026-W30",
    isOverLimit: false,
    percentUsed: 25,
  }))
}

export function checkQuotaBeforeRecord(charCount: number): Promise<boolean> {
  return invokeOrMock("check_quota_before_record", { charCount }, () => true)
}

export function adminSetUserQuota(userId: string, limit: number): Promise<void> {
  return invokeOrMock("admin_set_user_quota", { userId, limit }, () => undefined)
}

export function getWeeklyFreeChars(): Promise<number> {
  return invokeOrMock("get_weekly_free_chars", undefined, () => 10000)
}

File: openless-all/app/src-tauri/src/commands/billing.rs
//! Billing IPC commands — quota checking, usage recording, status queries.

use super::*;

#[tauri::command]
pub fn get_quota_status(coord: CoordinatorState<'_>) -> Result<crate::billing::QuotaStatus, String> {
    let user_id = coord
        .active_user_id()
        .ok_or("请先登录".to_string())?;
    coord.billing().get_quota_status(&user_id)
}

#[tauri::command]
pub fn check_quota_before_record(
    coord: CoordinatorState<'_>,
    char_count: u64,
) -> Result<bool, String> {
    let user_id = coord
        .active_user_id()
        .ok_or("请先登录".to_string())?;
    let action = coord.billing().check_quota(&user_id, char_count)?;
    Ok(matches!(action, crate::billing::QuotaAction::Allow))
}

#[tauri::command]
pub fn admin_set_user_quota(
    coord: CoordinatorState<'_>,
    user_id: String,
    limit: u64,
) -> Result<(), String> {
    // P0 修复：此前无任何鉴权，任意调用方可把任意用户配额改为无限。
    crate::commands::auth::require_admin(&coord)?;
    coord.billing().set_override_limit(&user_id, limit)
}

#[tauri::command]
pub fn get_weekly_free_chars() -> u64 {
    crate::billing::WEEKLY_FREE_CHARS
}
